contracts/exchangeIssuance/FlashMintWrapped.sol

```diff
@@ -17,220 +17,1087 @@
+   *
+   * @param _outputToken                Output token that will be bought
+   * @param _swapData                   ComponentSwapData (underlyingERC20 -> output token) for each _redeemComponents in the same order
+   * @param _invokeUnwrapData           ComponentInvokeWrapData (wrapped Set component -> underlyingERC20) for each _redeemComponents in the same order
+   * @param _redeemComponents           redemption components gotten from DebtIssuanceModule.getRequiredComponentRedemptionUnits()
+   * @param _redeemAmounts              redemption units gotten from DebtIssuanceModule.getRequiredComponentRedemptionUnits()
+   *
+   * @return totalOutputTokenObtained   total output token amount obtained
+   */
+  function _unwrapAndSwapComponents(
+    IERC20 _outputToken,
+    ComponentSwapData[] calldata _swapData,
+    ComponentInvokeWrapData[] calldata _invokeUnwrapData,
+    address[] memory _redeemComponents,
+    uint256[] memory _redeemAmounts
+  ) internal returns (uint256 totalOutputTokenObtained) {
+    // for each component in the swapData / wrapData / redeemComponents array:
+    // 1. unwrap from wrapped set component to unwrapped underlyingERC20 in swapData
+    // 2. swap from underlyingERC20 token to output token (exact from obtained underlyingERC20 amount)
+
+    for (uint256 i = 0; i < _redeemComponents.length; ++i) {
+      // default redeemed amount is maximum possible amount that was redeemed for this component
+      // this is recomputed if the redeemed amount is unwrapped to the actual unwrapped amount
+      uint256 redeemedAmount = _redeemAmounts[i];
+
+      // if the set component is the output token, no swapping or wrapping is needed
+      if (address(_outputToken) == _redeemComponents[i]) {
+        // add maximum possible amount that was redeemed for this component to totalOutputTokenObtained (=redeemedAmount)
+        totalOutputTokenObtained = totalOutputTokenObtained.add(redeemedAmount);
+        continue;
+      }
+
+      // transform wrapped token into unwrapped version (unless it's the same)
+      if (_swapData[i].underlyingERC20 != _redeemComponents[i]) {
+        // snapshot unwrapped balance before to compute the actual redeemed amount of underlying token (due to unknown exchange rate)
+        uint256 unwrappedBalanceBefore = IERC20(_swapData[i].underlyingERC20).balanceOf(
+          address(this)
+        );
+
+        _unwrapComponent(
+          _invokeUnwrapData[i],
+          _redeemComponents[i]
+        );
+
+        // recompute actual redeemed amount to the underlyingERC20 token amount obtained through unwrapping
+        uint256 unwrappedBalanceAfter = IERC20(_swapData[i].underlyingERC20).balanceOf(
+          address(this)
+        );
+        redeemedAmount = unwrappedBalanceAfter.sub(unwrappedBalanceBefore);
+      }
+
+      // swap redeemed and unwrapped component to output token
+      uint256 boughtOutputTokenAmount = _swapFromExact(
```

Comment:
Also here I think we need a check to make sure the user doesn't sell more underlying tokens than he has previously obtained in unwrapping.